contracts/Launchpad.sol

```diff
@@ -0,0 +1,492 @@
+        launchTokenAddress[launchId + 1] = launchTokenAddress[launchId];
+        launchBaseAllocation[launchId + 1] = launchBaseAllocation[launchId];
+
+        emit LaunchAdded(launchId + 1, 2);
+    }
+
+    function removeLaunch(uint256 launchId) external restricted {
+        delete launches[launchId];
+    }
+
+    // UPDATING DETAILS
+    function setLaunchDetails(uint256 launchId, uint256 tokenPrice, uint256 startTime, uint256 fundsToRaise) external restricted {
+        launchTokenPrice[launchId] = tokenPrice;
+        launchStartTime[launchId] = startTime;
+        launchFundsToRaise[launchId] = fundsToRaise;
+    }
+
+    function updateLaunchName(uint256 launchId, string calldata name) external restricted {
+        launches[launchId].name = name;
+    }
+
+    function updateLaunchTokenSymbol(uint256 launchId, string calldata tokenSymbol) external restricted {
+        launches[launchId].tokenSymbol = tokenSymbol;
+    }
+
+    function updateLaunchDetailsJsonUri(uint256 launchId, string calldata detailsJsonUri) external restricted {
+        launches[launchId].detailsJsonUri = detailsJsonUri;
+    }
+
+    function updateLaunchImageUrl(uint256 launchId, string calldata imageUrl) external restricted {
+        launches[launchId].imageUrl = imageUrl;
+    }
+
+    function updateLaunchFundingTokenAddress(uint256 launchId, address fundingTokenAddress) external restricted {
+        launches[launchId].fundingTokenAddress = fundingTokenAddress;
+    }
+
+    function updateLaunchTokenPrice(uint256 launchId, uint256 tokenPrice) external restricted {
+        launchTokenPrice[launchId] = tokenPrice;
+    }
+
+    function updateLaunchStartTime(uint256 launchId, uint256 startTime) external restricted {
+        require(launchTokenPrice[launchId] != 0 && launchFundsToRaise[launchId] != 0, "Set token price and funds to raise first");
+        launchStartTime[launchId] = startTime;
+    }
+
+    function updateLaunchFundsToRaise(uint256 launchId, uint256 fundsToRaise) external restricted {
+        launchFundsToRaise[launchId] = fundsToRaise;
+    }
+
+    function updateLaunchTokenAddress(uint256 launchId, address tokenAddress) external onlyPhase1(launchId) restricted {
+        launchTokenAddress[launchId] = tokenAddress;
+        launchTokenAddress[launchId + 1] = tokenAddress;
+    }
+
+    // VESTING
+
+    function enablePeriodicVesting(uint256 launchId, uint256 percentage) external onlyPhase1(launchId) isTokenAddressSet(launchId) restricted {
+        require(launchLinearVestingsDurations[launchId] == 0, "Linear vesting enabled");
+        require(getTotalUnlockedPercentage(launchId) + percentage <= 100, "Total percentage exceeded");
+        uint256 totalLaunchRaised = launchTotalRaised[launchId].add(launchTotalRaised[launchId + 1]);
+        uint256 contractBalance = IERC20(launchTokenAddress[launchId]).balanceOf(address(this));
+        uint256 decimals = ERC20(launchTokenAddress[launchId]).decimals();
+        if(decimals > 18) {
```

Comment:
Seems similar to the method you did in common.sol, maybe we can use it here as well?